root/components/FieldErrors.js

```diff
@@ -11,23 +11,34 @@ import React from 'react';
 
 import subfieldErrors, {type FieldShape} from '../utility/subfieldErrors';
 
-const buildErrorListItem = (error, index) => (
-  <li key={index}>{error}</li>
-);
+const buildErrorListItem = (error, hasHtmlErrors, index) => {
+  if (hasHtmlErrors) {
+    return (
+      <li
+        dangerouslySetInnerHTML={{__html: error}}
```

Comment:
It is vulnerable to XSS injection via entity’s name, e.g.
```javascript
</a><script>alert("XSS")</script><a>
```

Reply:
I'm surprised if `Translation::expand` doesn't HTML-escape the name there. That's a pretty glaring bug which we should address.

But you should also be able to use `expand2react` in place of `dangerouslySetInnerHTML` here, and that does seem safer overall.

Reply:
Steps to reproduce:
* In a release group, create a new release duplicating an existing release but the recordings associated with the two first tracks have to be swapped in the Recordings tab
* Rename one of these recordings `</a><script>alert("XSS")</script><a>`
* Merge releases and choose strategy “Merge mediums and recordings”

Reply:
expand2react seems to be affected as well, unless I'm missing something.

Reply:
This should be fixed with 363b3efa5235e2191a8e0dc6529906dff6a2ede7 on TT, not sure if that will cause issues once it moves to React though?

Reply:
> expand2react seems to be affected as well, unless I'm missing something.

`expand2react` should only allow tags from a [short list of HTML tags](https://github.com/metabrainz/musicbrainz-server/blob/v-2020-01-20/root/static/scripts/common/i18n/expand2react.js#L49). Doesn’t it? That should prevent `script` tag at least, not sure about roguish links though.

Reply:
> This should be fixed with [363b3ef](https://github.com/metabrainz/musicbrainz-server/commit/363b3efa5235e2191a8e0dc6529906dff6a2ede7) on TT, not sure if that will cause issues once it moves to React though?

Just tested, it works perfectly for merge error messages at least. It should not have any side-effect as `_link_recording` is only called for merge error messages, not otherwise.

But I would still recommend to remove that `dangerouslySetInnerHTML` if possible.

Reply:
I did change that to expand2react :) (and yeah, sorry, I was being stupid when I posted about expand2react earlier - I forgot I'm not actually using React on the page with the errors and only changed the file to keep parity!

Reply:
Alright, I only checked this conversation and the added commit, not the updated one, thanks!

Reply:
The added commit https://github.com/metabrainz/musicbrainz-server/pull/1361/commits/363b3efa5235e2191a8e0dc6529906dff6a2ede7 is working around a bug in `Translation::expand` as @mwiencek suggested. This bug might affect pages using the TT macro `display_relationship` too.

Since TT macros will be replaced one day, it would be nice to stop using `Translation::expand` here already.

Reply:
The current code in the PR looks reasonable to me. I do agree that the right thing for `Translation::expand` to do would be to escape the name itself, but I'm afraid we might abuse that in some places to pass HTML inside the link, or are already escaping the HTML outside of the call in some places (so we'd end up double-escaping things by changing it).

Reply:
> The current code in the PR looks reasonable to me.

It works flawlessly but it also prepares for a future vulnerability as any added `message` in Perl code will have to be checked for correctly encoding entities. Ideally, this task should not be carried out by the data layer.

> I do agree that the right thing for `Translation::expand` to do would be to escape the name itself, but…

I thought it would be to pass fully expandable messages rather than HTML messages from Perl to React/TT and let it deal with encoding entities. Does it seem reasonable? Might be a separate PR.

Reply:
> I thought it would be to pass fully expandable messages rather than HTML messages from Perl to React/TT and let it deal with encoding entities. Does it seem reasonable?

Well, you'd have to pass back the interpolation arguments too (in this case, recording JSON). So the error message would have to be a JSON document that you parse and pass to `l`/`ln`/`lp`.

This might work with React, but root/release/merge.tt doesn't use React for these error messages. I don't quite understand how you envision it working for TT, since we have to generate HTML at some point in that case.

Reply:
Created MBS-10632 for follow-up.